[AArch64][Build Attributes] Improve Parsing and Formatting

sivan-shani opened the pull request:
- Removed assertion for duplicate values as adding them is valid.
- Fix parsing: reject strings for unknown tags, allow any value for Tag_PAuth_Platform and Tag_PAuth_Schema.
- Print tags by using numbers with comments to reduce compiler-assembler dependencies.
- Parsing error messages now only point to the symbol (^) instead of printing it.

ostannard commented:
Would you mind splitting this into multiple patches to make it easier to review? Based on the PR description, this looks like it could be three patches:
* Improve testing of the existing behaviour, no functional changes
* Do not allow unknown tags to be strings, and add/change tests for this
* Remove assertion when value exists, and add/change tests for this

I think these are closely related enough to stay in one PR, but splitting them into patches allows me (and other reviewers) to see which test changes correspond to which code changes.

sivan-shani commented:
Since it is a relatively small patch, I will try to describe what has changed and what belongs where.
If it is still not clear. please let me know and I will try to split the PR to patches.

**Tag_PAuth_Platform and Tag_PAuth_Schema**
Allow assembly to pass any unsigned values, not only 0 or 1
code: _llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp_
Achieved by changing the condition
`if (ActiveSubsectionID == AArch64BuildAttrs::AEABI_FEATURE_AND_BITS) {`
test: _llvm/test/MC/AArch64/aarch64-build-attributes-asm-aeabi-aeabi-known.s_
```
.aeabi_attribute Tag_PAuth_Platform, 7
.aeabi_attribute Tag_PAuth_Schema, 777
```
also removing `.aeabi_attribute Tag_PAuth_Platform, 4` from 
_llvm/test/MC/AArch64/aarch64-build-attributes-asm-aeabi-err-attrs.s_

**Do not allow unknown tags to be strings**
code: _llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp_
`case AArch64BuildAttrs::VENDOR_UNKNOWN:` under `} else if (Parser.getTok().is(AsmToken::Identifier)) {`
now gives an error.
also moved `if (Parser.getTok().is(AsmToken::Integer)) {` before so it is clear that the integer case is being handled.
test: _llvm/test/MC/AArch64/aarch64-build-attributes-asm-non_aeabi-err.s_
`.aeabi_attribute str_not_int, "1"`

**Remove assertion when value exists**
_llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp_
```
// Case Item.IntValue == Value, no need to emit twice
assert(0 &&
"AArch64 build attribute: An attribute with the same tag "
"and a same value already exists");
```
removed. This is not an error and now release and debug behave the same.
test: _llvm/test/MC/AArch64/aarch64-build-attributes-asm-aeabi-aeabi-known.s_
```
.aeabi_attribute Tag_PAuth_Schema, 777
.aeabi_attribute Tag_PAuth_Schema, 777
```
produce same object as before.

**general testing (and other) improvement**
-> error msgs at _llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp_ standardize, made 2 specific cases like the others.
-> change files names, 'private' -> 'non-aeabi', more clear.
-> change files structure - put syntax and logical tests separately.
-> add file mixing 'aeabi' and 'non-aeabi' tests _llvm/test/MC/AArch64/aarch64-build-attributes-asm-aeabi-mixed.s_

ostannard commented:
You now have 5 different changes in one patch, please split it up.

sivan-shani committed:
[AArch64][Build Attributes] Improve testing (delete previous tests files)

This commit deletes previous tests files, following commits adding new test files.

sivan-shani committed:
[AArch64][Build Attributes] Standardize parsing error messages

Parsing error message should not print the symbol triggering error.
(Only point to it with the symbol '^')

sivan-shani committed:
[AArch64][Build Attributes] Print Tags as Numbers with Comments

Print Tags as numbers followed by comments instead of using string representations.
For example:
.aeabi_attribute 0, 1 @ Tag_Feature_BTI

instead of:
.aeabi_attribute Tag_Feature_BTI, 1

This reduces dependencies between compiler and assembler versions.

sivan-shani committed:
[AArch64][Build Attributes] Fix Parsing of Some Tags and Values

- Reject strings for unknown tags.
- Allow any value for Tag_PAuth_Platform and Tag_PAuth_Schema.

sivan-shani committed:
[AArch64][Build Attributes] Remove Assertion for Duplicate Values

Adding the same value twice is valid, so the assertion has been removed.

sivan-shani committed:
[AArch64][Build Attributes] Add test files

Add test files

sivan-shani commented:
Break down the change to several patches

ostannard reviewed (commented):
When splitting patches up, please make sure that the tests are updated in the same patch which makes the functional change. This makes it much easier for the reviewer to see what has/hasn't been tested, and the tests are also useful to the reviewer as an example of what behaviour is changed/added by the patch. In the current version of this PR, you've deleted most of the tests in the first commit and re-added them in the last one, which makes it harder to see what has been added/changed in them.

sivan-shani committed:
[AArch64][Build Attributes] Same Tag different Value: override

When a new value is provided for an existing tag, override.

sivan-shani commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp:
```diff
@@ -222,13 +222,9 @@ void AArch64TargetStreamer::emitAttribute(StringRef VendorName, unsigned Tag,
-                     "and a same value already exists");
-              return;
             }
+            // Case Item.IntValue == Value is permited.
```
- Same tag, same value: ignore
Test- _aarch64-build-attributes-asm-aeabi-aeabi-known.s_
```
.aeabi_attribute Tag_PAuth_Schema, 777
.aeabi_attribute Tag_PAuth_Schema, 777
```
- Same tag, different value: override
Test- _aarch64-build-attributes-asm-aeabi-mixed.s_ (just been added)

sivan-shani committed:
[AArch64][Build Attributes] Improve Parsing and Formatting

AArch64 assembly comments should be '//'
Lexer remove comments, no need to handle them when parsing.

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp:
```diff
@@ -214,22 +215,20 @@ void AArch64TargetStreamer::emitAttribute(StringRef VendorName, unsigned Tag,
       for (MCELFStreamer::AttributeItem &Item : SubSection.Content) {
+        // Tag already exists
         if (Item.Tag == Tag) {
-          if (!Override) {
```
The `Override` parameter is now unused.

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7948,7 +7947,9 @@ bool AArch64AsmParser::parseDirectiveAeabiSubSectionHeader(SMLoc L) {
   Parser.Lex();
-  // Parsing finished, check for trailing tokens.
+
+  // Parsing finished. Check for trailing characters (no need to check for
```
The extra comment isn't needed, this is true everywhere in the assembly parser.

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -7986,14 +7987,22 @@ bool AArch64AsmParser::parseDirectiveAeabiAArch64Attr(SMLoc L) {
     TagStr = Parser.getTok().getIdentifier();
     switch (ActiveSubsectionID) {
     default:
+      // Should not happen
```
This comment isn't adding anything useful. It would actually be better to remove the default case, since we expect this enum to be fully covered: https://llvm.org/docs/CodingStandards.html#don-t-use-default-labels-in-fully-covered-switches-over-enumerations

ostannard commented on llvm/lib/Target/AArch64/AsmParser/AArch64AsmParser.cpp:
```diff
@@ -8071,7 +8077,9 @@ bool AArch64AsmParser::parseDirectiveAeabiAArch64Attr(SMLoc L) {
   Parser.Lex();
-  // Parsing finished, check for trailing tokens.
+
+  // Parsing finished. Check for trailing characters (no need to check for
```
The extra comment isn't needed, this is true everywhere in the assembly parser.

ostannard commented on llvm/lib/Target/AArch64/MCTargetDesc/AArch64TargetStreamer.cpp:
```diff
@@ -214,22 +215,20 @@ void AArch64TargetStreamer::emitAttribute(StringRef VendorName, unsigned Tag,
-              return;
-            }
+          // New value for existing tag: update tag
+          if ((unsigned(-1) != Value && Item.IntValue != Value) ||
```
Why do you need to check whether the value is different to the previous one? This will have the same effect as always updating it.

sivan-shani replied:
It might eliminate unnecessary operations, depending on which is preferable: the comparison or updating the values.
It is reasonable to believe that the comparison is cheaper. Since neither case makes sense and should not happen unless as a result of a mistake or oversight, there is no reason to assume that the case requiring updates will be more frequent then the case which does not. Therefore, adding the check makes sense.

Additionally, it enhances readability and makes future behavior changes easier if needed.

ostannard replied:
As you point out, this is very cold code, so I don't think we should be worrying about performance here.

I disagree that this increases readability - it's just adding extra code which doesn't do anything, and nobody reading this is going to be surprised y an assignment appearing to have no effect if the two values are the same.

sivan-shani committed:
[AArch64][Build Attributes] Remove unused 'Override' parameter

Since the function accepts assembly that adds a different value for an existing tag,
the 'Override' parameter is no longer needed.

sivan-shani committed:
[AArch64][Build Attributes] Remove redundant comments and unneeded case

- Remove redundant comments
- Remove default case from a fully covered switch

MaskRay commented on llvm/test/MC/AArch64/aarch64-build-attributes-asm-aeabi-aeabi-known.s:
```diff
@@ -0,0 +1,43 @@
+// RUN: llvm-mc -triple=aarch64 %s -o - | FileCheck %s --check-prefix=ASM
```
For new tests just use `MC/AArch64/xxx` instead of `MC/AArch64/aarch64-xxx`.

MaskRay commented on llvm/test/MC/AArch64/aarch64-build-attributes-asm-aeabi-mixed.s:
```diff
@@ -0,0 +1,50 @@
+// RUN: llvm-mc -triple=aarch64 -filetype=obj %s -o - | llvm-readelf --hex-dump=.ARM.attributes - | FileCheck %s --check-prefix=ELF
+
+// ASM: .aeabi_subsection	subsection_a, optional, uleb128
+// ASM: .aeabi_subsection	aeabi_subsection, optional, ntbs
```
Use `-NEXT:` whenever applicable

sivan-shani replied:
Assembly lines that does not contain a comment emit an extra new line.
Hence some line has an extra new line, and the directive '-NEXT' can not be used.
The cause seems to be `llvm/lib/MC/MCAsmStreamer.cpp`
Assembly lines that do not contain a comment emit an extra newline.
As a result, some lines have an extra newline, making the -NEXT directive unusable.
The issue appears to originate from llvm/lib/MC/MCAsmStreamer.cpp.
"""
  inline void EmitEOL() {
    // Dump Explicit Comments here.
    emitExplicitComments();
    // If we don't have any comments, just emit a \n.
    if (!IsVerboseAsm) {
      OS << '\n';
      return;
    }
    EmitCommentsAndEOL();
  }
  
  void MCAsmStreamer::EmitCommentsAndEOL() {
  if (CommentToEmit.empty() && CommentStream.GetNumBytesInBuffer() == 0) {
    OS << '\n';
    return;
  }
  ...
"""
Removing the extra newline causes a large number of other tests to fail, so it has been left unchanged for now.

ostannard replied:
Would it help to use that mechanism, and add the comment with `addExplicitComment()` instead of a literal "//" string?

sivan-shani replied:
Can not find a decent way to the this within the way `AArch64TargetAsmStreamer` is implemented. Leaving it the way it is.

ostannard replied:
What is the difference between `AArch64TargetAsmStreamer` and, for example, the ARM equivalent, where the `.eabi_attribute` directive emits a comment without any spurious newlines?

Does this occur when generating code from IR (using llc or clang), or only when parsing assembly? That might help narrow down which code path the bug is in.

sivan-shani replied:
The different between Arm and AArch64 might be due to the way `MCAsmStreamer::addExplicitComment` handle the different comments. It has a specific implementation for `\\` comments then `@` comments.

ostannard replied:
Ok, this only seems to affect assembly->assembly, which I think is only used for testing, so it's a minor issue and doesn't need to block the rest of this PR.

sivan-shani committed:
Remove prefix aarch64 from test files residing in the AArch64 folder

sivan-shani committed:
For llvm-mc, omit -o - (default)

sivan-shani committed:
handle unnecessary switch defaults

sivan-shani committed:
Remove unnecessary condition check when tag exists.

ostannard reviewed (approved):
LGTM
